app/src/main/java/org/nieghborhoodbikeworks/nbw/ui/queue/QueueFragment.java

```diff
@@ -1,114 +1,224 @@
+            mQueueDatabase.updateChildren(childUpdates);
+            updateQueue();
+            FirebaseAuth.getInstance().signOut();
+            mUser = null;
+        } catch (NullPointerException e) {
+            // Redirect the user to the login screen if they aren't signed in and would
+            // like to be added to, or removed from, the queue
+            Navigation.findNavController(view).navigate(R.id.loginFragment);
+        }
+    }
+
+    private void dequeueUser() {
+        // If there is no user currently signed in, mUser.getName() will throw a NPE
+        try {
+            mQueueDatabase.child(mUser.getName()).removeValue();
+            updateQueue();
+            FirebaseAuth.getInstance().signOut();
+            mUser = null;
+        } catch (NullPointerException e) {
+            // Redirect the user to the login screen if they aren't signed in and would
+            // like to be added to, or removed from, the queue
+            Navigation.findNavController(view).navigate(R.id.loginFragment);
+        }
+    }
+
+    /**
+     * The updateQueue method reads from the queue node in the database and adds the users to the mQueue
+     * ArrayList.
+     */
+    private void updateQueue() {
+        // Reads the data on the "queue" node in the database
+        ValueEventListener queueListener = new ValueEventListener() {
+            @Override
+            public void onDataChange(DataSnapshot dataSnapshot1) {
+                ArrayList<String> updatedQueue = new ArrayList<>();
+                // Iterates through the nodes in the queue
+                for (DataSnapshot ds1 : dataSnapshot1.getChildren()) {
+                    // key = mUser.getName(), value = mUser
+                    String key = ds1.getKey();
+                    // Add users in the queue
+                    if(!key.equals("empty")) {
+                        updatedQueue.add(key);
                     }
-                })
-                .setNegativeButton("No", new DialogInterface.OnClickListener() {
-                    public void onClick(DialogInterface dialog, int which) {
-//                      some logic for when the user decides to not be added to the queue
+                }
+
+                int i = 0;
+                boolean userDequeued = false;
+
+                if (mQueue.size() > updatedQueue.size()) {
+                    for (String user : updatedQueue) {
+                        if (!(user.equals(mQueue.get(i)))) {
+                            userDequeued = true;
+                            break;
+                        }
+                        i++;
+                    }
+                }
+
+                mQueue = updatedQueue;
+                if (userDequeued) {
+                    mAdapter.notifyItemRemoved(i);
+                }
+                //TODO: Find a better way to update the RecyclerView Adapter
+                mAdapter = new QueueAdapter(getActivity(), mQueue);
```

Comment:
If you decide to push your database access code into the viewmodel, be careful and make sure that this adapter doesn't end up in the viewmodel as well. It contains a reference to the view, not best practice.